from snakemake.utils import min_version

##### set minimum snakemake version #####
min_version("6.4.1")

# this container defines the underlying OS for each job when using the workflow
# with --use-conda --use-singularity
#container: "docker://continuumio/miniconda3"
# workdir: config.get("workdir")

#define some global variables
base_ref=config.get('paths').get('base_ref') + "/" + config["ref"]["species"] + "/" + config["ref"]["build"] + "/" + str(config["ref"]["release"]) 
ref_genome=config["ref"]["species"] + "." + config["ref"]["build"] + "." + str(config["ref"]["release"]) + ".fasta"
ref_annots=config["ref"]["species"] + "." + config["ref"]["build"] + "." + str(config["ref"]["release"]) + ".gtf"

##### load rules #####
include: "rules/common.smk"
include: "rules/ref.smk"
# include: "rules/trim.smk"
include: "rules/align.smk"
# include: "rules/qc.smk"
# include: "rules/diffexp.smk"
# include: "rules/countmatrix.smk"
# include: "rules/genotyping.smk"


##### target rules #####

rule all:
    input:
        # "resources/genome.fasta.fai",
        base_ref + "/" + ref_genome+".fai",
        expand(base_ref + "/" + ref_genome+".{ext}", ext=["amb", "ann", "bwt", "pac", "sa"]),
        directory(base_ref + "/star_genome")
        # "results/pca.pdf",
        # get_final_output(),
        # "results/counts/all_tpm.tsv",
        # "results/qc/multiqc_report.html",
